onnx/onnx.in.proto

```diff
@@ -196,6 +204,93 @@ message NodeProto {
+//      NodeProto.op_type == "Graph" and
+//      NodeProto.name == ModelProto.graph.name.
+//   2. Initializers in ModelProto.graph.initializer can be referenced by
+//      nodes in TrainingInfoProto.algorithm.nodes and
+//      TrainingInfoProto.initialization.nodes. It implies that training
+//      nodes cannot produce values with names identical to inference
+//      initializers.
+message TrainingInfoProto {
+  // This field describes a graph to compute the initial tensors
+  // upon starting the training process. Initialization graph has no input
+  // and can have multiple outputs. Assigning those outputs to the associated
+  // initializers follows the relation defined by update_binding as if
+  // TrainingInfoProto.initialization is another training algorithm. Usually,
+  // trainable tensors in neural networks are randomly initialized.
+  // To achieve that, for each tensor, the user can put a random operator such
+  // as RandomNormal or RandomUniform in TrainingInfoProto.initialization.node
+  // and assign its random output to the specific tensor using "update_binding".
+  //
+  // By default, this field is an empty graph and its evaluation does not
+  // produce any output.
+  optional GraphProto initialization = 1;
+
+  // This field represents a training algorithm step. Given required inputs,
+  // it computes outputs to update initializers in either its own initializer
+  // list or in ModelProto.graph.initializer. In general, this graph contains
+  // loss node, gradient node, optimizer node, increment of iteration count,
+  // and calls to the inference graph.
+  //
+  // The field algorithm.node can invoke ModelProto.graph as a node.
+  // The initializers defined in ModelProto.graph.initializer are visible in
+  // the "algorithm".
+  //
+  // The field MUST be present.
+  optional GraphProto algorithm = 2;
+
+  // Gradient-based training is usually an iterative procedure. In one gradient
+  // descent iteration, we apply
+  //
+  // x = x - r * g
+  //
+  // where "x" is the optimized tensor, "r" stands for learning rate, and "g" is
+  // gradient of "x" with respect to a chosen loss. To avoid adding assignments
+  // into the training graph, the
+  // user needs to save "y = x - r * g" into a TrainingInfoProto.algorithm. To
+  // tell runtime that "y" should be assigned to "x", the field
+  // "update_binding" may contain a pair of strings, "y"
+  // (key of StringStringEntryProto) and "x" (value of StringStringEntryProto).
+  // Similarly, for a neural network which contains multiple tensors, there can
+  // be multiple key-value pairs in "update_binding".
+  //
+  // In general, the user can use "update_binding" to assign any output of
+  // "algorithm" and "initialization" to any initializer in
+  // ModelProto.graph.initializer and TrainingInfoProto.algorithm.initializer.
+  // One initializer can only be assigned up to once; it implies that the
+  // values in "update_binding" must be unique.
```

Comment:
A couple of related points: (a) Should we have a separate update_binding for initialization and a separate one for algorithm? Otherwise, we need to be clear whether we expect the two graphs to use same output-names for one given initializer, or allow both (init_x, x) and (update_x, x) in the bindings. (b) Should we swap the order of names in the update_bindings? That has two advantages: it is more consistent with the ordering in an assignment statement. It also clarifies that only one update-value should be specified for an initializer (since the first entry is the "key").

Reply:
(a) My understanding is that we allow both (init_x, x) and (update_x, x) in `update_binding`. It means `update_binding` is an union of assignments from initialization and training.

(b) I personally feel `(key, value)=(W_new, W)` is closer to what will happen in the runtime. After finishing the computation of `algorithm` and produce an output `W_new`, we need a mapping from `W_new` to the updated initializer. If we store `(key, value)=(W, W_new`), we will need to convert '(W, W_new)` to `(W_new, W)` when creating output-to-initializer mapping.

Reply:
Re. (a): I think splitting the two simplifies the semantics. Re. (b) also: I think we should think in terms of defining the semantics formally (using appropriate programming/mathematical notation). My opinion is that difference in this sense is not much, but the order (lhs, rhs) is common for an assignment and similarly (variable, value) is common for a binding in functional languages.

Mathematically, the underlying state is a "map" M that maps tensor-variables to their values. The execution of a single update-binding (W, W_new) updates M to M[W => W_new].

Reply:
Re Re (a): Ok. I will create `initialization_binding` and `algorithm_binding` and remove `update_binding`.

Re Re (b): Ok. I will swap keys and values.